Translation of the "Handling Events" page (#3)
reactjs/pt-BR.reactjs.org: content/docs/handling-events.md

---
id: handling-events
title: Manipulando eventos
permalink: docs/handling-events.html
prev: state-and-lifecycle.html
next: conditional-rendering.html
redirect_from:

>   - "docs/events-ko-KR.html"

Reviewer: cc @tesseralis what is expected here? `"docs/events-pt-BR.html"`?

>   - "docs/events-ko-KR.html"

Author: I had not seen. I do not know how to use redirect_from, I put pt-BR?

>   - "docs/events-ko-KR.html"

Reviewer: let's skip this for now and follow-up if something is not correct. resolving by now.

---

> Manipular eventos em elementos com React é muito semelhante ao tratamento de eventos em elementos no DOM. Existem algumas diferenças sintáticas:

Reviewer 2: ```suggestion
Manipular eventos em elementos React é muito semelhante a manipular eventos em elementos do DOM. Existem algumas diferenças sintáticas:
```

* Eventos em React são nomeados usando camelCase ao invés de letras minúsculas.
* Com o JSX você passa uma função como manipulador de eventos ao invés de um texto.

> For exemplo, com HTML:

Reviewer 2: Alterar **for** para **por**

Reviewer 3: ```suggestion
Por exemplo, com HTML:
```

```html
<button onclick="activateLasers()">
  Ativar lasers
</button>
```

É ligeiramente diferente com React:

```js{1}
<button onClick={activateLasers}>
  Ativar lasers
</button>
```

Outra diferença é que você não pode retornar `false` para evitar o comportamento padrão no React. Você deve chamar `preventDefault` explícitamente. Por exemplo, com HTML simples, para evitar que um link abra uma nova página, você pode escrever:

```html
<a href="#" onclick="console.log('O link foi clicado.'); return false">

>   Clique em mim

Reviewer 3: ```suggestion
Clique Aqui
```

</a>
```

> Em React, isso poderia ser:

Reviewer 2: ```suggestion
No React, isso poderia ser:
```

```js{2-5,8}
function ActionLink() {
  function handleClick(e) {
    e.preventDefault();
    console.log('O link foi clicado.');
  }

  return (
    <a href="#" onClick={handleClick}>

>       Clique em mim

Reviewer 3: ```suggestion
Clique Aqui
```

    </a>
  );
}
```

> Aqui, "`e`" é um synthetic event. O React define esses eventos sintéticos de acordo com a [especificação W3C](https://www.w3.org/TR/DOM-Level-3-Events/). Então, não precisa se preocupar com a compatibilidade entre navegadores. Veja a página [`SyntheticEvent`](/docs/events.html) para saber mais.

Reviewer 2: ```suggestion
Aqui, "`e`" é um synthetic event. O React define esses eventos sintéticos de acordo com a [especificação W3C](https://www.w3.org/TR/DOM-Level-3-Events/). Então, não precisamos nos preocupar com a compatibilidade entre navegadores. Veja a página [`SyntheticEvent`](/docs/events.html) para saber mais.
```

> Aqui, "`e`" é um evento sintético. O React define esses eventos sintéticos de acordo com a [especificação W3C](https://www.w3.org/TR/DOM-Level-3-Events/), então, não precisa se preocupar com a compatibilidade entre navegadores. Veja a página [`SyntheticEvent`](/docs/events.html) para saber mais.

Reviewer 4: Is this case `SyntheticEvent` will refer to a page, wouldn't be better translating this to use the Portuguese? Because the name will match with the Portuguese page's title

Reviewer 2: I agree with @eduardomoroni

> Aqui, "`e`" é um evento sintético. O React define esses eventos sintéticos de acordo com a [especificação W3C](https://www.w3.org/TR/DOM-Level-3-Events/), então, não precisa se preocupar com a compatibilidade entre navegadores. Veja a página [`SyntheticEvent`](/docs/events.html) para saber mais.

Reviewer 3: ```suggestion
Aqui, "`e`" é um synthetic event. O React define esses eventos sintéticos de acordo com a [especificação W3C](https://www.w3.org/TR/DOM-Level-3-Events/). Então, não precisa se preocupar com a compatibilidade entre navegadores. Veja a página [`SyntheticEvent`](/docs/events.html) para saber mais.
```

Ao usar o React você geralmente não precisa chamar `addEventListener` para adicionar ouvintes a um elemento no DOM depois que ele é criado. Ao invés disso você pode apenas definir um ouvinte quando o elemento é inicialmente renderizado.

> Quando você define um componente usando uma [classe do ES6](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes), um padrão comum é que um manipulador de eventos seja um método na classe. Por exemplo, este componente `Toggle` renderiza um botão que permite ao usuário alternar entre os estados "ON" e "OFF":

Reviewer 2: It would be nice to point the links from MDN to the pt-BR docs

Author: Ah sim, eu não olhei os links ai não vi que tinha que trocar.

> Quando você define um componente usando uma [classe do ES6](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes), um padrão comum é que um manipulador de eventos seja um método na classe. Por exemplo, este componente `Toggle` renderiza um botão que permite ao usuário alternar entre os estados "ON" e "OFF":

Reviewer 3: ```suggestion
Quando você define um componente usando uma [classe do ES6](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Classes), um padrão comum é que um manipulador de eventos seja um método na classe. Por exemplo, este componente `Toggle` renderiza um botão que permite ao usuário alternar entre os estados "ON" e "OFF":
```

);
```

[**Experimente no CodePen**](http://codepen.io/gaearon/pen/xEmzGg?editors=0010)

> Você precisa ter cuidado com o significado do `this` nos callbacks do JSX. Em JavaScript, os métodos de classe não são [vinculados](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_objects/Function/bind) por padrão. Se você esquecer de fazer o bind de `this.handleClick` e passá-lo para um `onClick`, o `this` será `undefined` quando a função for realmente chamada.

Reviewer 2: It would be nice to point the links from MDN to the pt-BR docs

> Você precisa ter cuidado com o significado do `this` nos callbacks do JSX. Em JavaScript, os métodos de classe não são [vinculados](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_objects/Function/bind) por padrão. Se você esquecer de fazer o bind de `this.handleClick` e passá-lo para um `onClick`, o `this` será `undefined` quando a função for realmente chamada.

Reviewer 3: ```suggestion
Você precisa ter cuidado com o significado do `this` nos callbacks do JSX. Em JavaScript, os métodos de classe não são [vinculados](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_objects/Function/bind) por padrão. Se você esquecer de fazer o bind de `this.handleClick` e passá-lo para um `onClick`, o `this` será `undefined` quando a função for realmente chamada.
```

> Este não é um comportamento específico do React; é uma parte de [como funcionam as funções em JavaScript](https://www.smashingmagazine.com/2014/01/understanding-javascript-function-prototype-bind/). Geralmente, se você referir a um método sem `()` depois dele, como `onClick={this.handleClick}`, você deve fazer o bind manual deste método.

Reviewer 3: ```suggestion
Este não é um comportamento específico do React. É uma parte de [como funcionam as funções em JavaScript](https://www.smashingmagazine.com/2014/01/understanding-javascript-function-prototype-bind/). Geralmente, se você referir a um método sem `()` depois dele, como `onClick={this.handleClick}`, você deve fazer o bind deste método.
```

Se ficar chamando "bind" incomoda você, há duas maneiras de contornar isso. Se você estiver usando a [sintaxe experimental de campos de classe pública](https://babeljs.io/docs/plugins/transform-class-properties/), você pode usar campos de classe para vincular callbaks corretamente:

```js{2-6}
class LoggingButton extends React.Component {

>   // Essa sintaxe garante que o `this` seja vinculado ao handleClick.

Reviewer 4: I'd rather don't translating some technical words, but that's not the case. Sometimes during this page, we use `bind` in English and other times we translate the word binding. What about or translation every `bind` term? Maybe we'll need to introduce that `bind` means `vincular`. I'm thinking about people with no English knowledge. This can confuse their minds.

Reviewer 4: https://github.com/reactjs/pt-BR.reactjs.org/issues/4#issuecomment-461766907

Reviewer 3: @eduardomoroni vou responder em português: quando a gente ver que o 'bind' no contexto da frase se refere ao método, não iremos traduzir o bind. Se não, traduziremos. Será uma questão de entender o contexto da frase. Ex.: [MDN Docs - Bind](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_objects/Function/bind)

Author: Então eu deixo do jeito que está ou eu corrijo?

Reviewer 5: > Esse bind é necessário para fazer o `this` funcionar dentro do callback

> Aqui utilizamos o `bind` para que o `this` funcione dentro da nossa callback

Acredito que não precise seguir a tradução ao pé da letra, escrever dessa forma mantem o uso da nomeclatura da função e deixa claro o motivo dela

Reviewer: I'll set this review as resolved as I think it is. please rollback if I'm mistaken

  // Atenção: essa é uma sintaxe *experimental*.
  handleClick = () => {
    console.log('this is:', this);
  }

  render() {
    return (
      <button onClick={this.handleClick}>

>         Clique em mim

Reviewer 3: ```suggestion
Clique Aqui
```

      </button>
    );
  }
}
```

Essa sintaxe é habilitada por padrão no [Create React App](https://github.com/facebookincubator/create-react-app).

> Se você não estiver usando a sintaxe de campos de classe, poderá usar uma [arrow function](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions) no retorno da chamada:

Reviewer 2: It would be nice to point the links from MDN to the pt-BR docs

> Se você não estiver usando a sintaxe de campos de classe, poderá usar uma [arrow function](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions) no retorno da chamada:

Reviewer 3: ```suggestion
Se você não estiver usando a sintaxe de campos de classe, poderá usar uma [arrow function](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Functions/Arrow_functions) no retorno da chamada:
```

}
```

> O problema com esta sintaxe é que um callback diferente é criado toda vez que o `LoggingButton` é renderizado. Na maioria dos casos, tudo bem. No entanto, se esse callback for passado para componentes inferiores através das propriedades, esses componentes poderão fazer uma renderização extra. Geralmente recomendamos a vinculação no construtor ou a sintaxe dos campos de classe para evitar esse tipo de problema de desempenho.

Reviewer 2: Please don't translate `props`, it could be confusing

Reviewer 3: ```suggestion
O problema com esta sintaxe é que um callback diferente é criado toda vez que o `LoggingButton` é renderizado. Na maioria dos casos, tudo bem. No entanto, se esse callback for passado para componentes inferiores através de props, esses componentes poderão fazer uma renderização extra. Geralmente recomendamos a vinculação no construtor ou a sintaxe dos campos de classe para evitar esse tipo de problema de desempenho.
```

> ## Passando argumentos para manipuladores de eventos {#passing-arguments-to-event-handlers}

Reviewer 3: ```suggestion
Passando Argumentos para Manipuladores de Eventos
```

Author: @glaucia86 Should I translate the link for this title?
```suggestion
## Passando Argumentos para Manipuladores de Eventos {#passando-argumentos-para-manipuladores-de-eventos}
```

Reviewer 3: @httpiago yes.

> Dentro de uma repetição é comum querer passar um parâmetro extra para um manipulador de evento. Por exemplo, se `id` é o ID de identificação da linha, qualquer um dos dois a seguir funcionará:

Reviewer 3: ```suggestion
Dentro de uma estrutura de repetição é comum querer passar um parâmetro extra para um manipulador de evento. Por exemplo, se `id` é o ID de identificação da linha, qualquer um dos dois a seguir funcionará:
```

```js
<button onClick={(e) => this.deleteRow(id, e)}>Deletar linha</button>
<button onClick={this.deleteRow.bind(this, id)}>Deletar linha</button>
```

> As duas linhas acima são equivalentes e usam [arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) e [`Function.prototype.bind`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind) respectivamente.

Reviewer 2: It would be nice to point the links from MDN to the pt-BR docs

Reviewer 3: ```suggestion
As duas linhas acima são equivalentes e usam [arrow functions](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Functions/Arrow_functions) e [`Function.prototype.bind`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_objects/Function/bind) respectivamente.
```

> Em ambos os casos, o argumento `e` representando o evento do React será passado como segundo argumento após o ID. Com uma arrow function, nós temos que passá-lo explicitamente, mas com o `bind` outros argumentos adicionais são automaticamente encaminhados.

Reviewer 3: ```suggestion
Em ambos os casos, o argumento `e` representando o evento do React será passado como segundo argumento após o ID. Com uma arrow function, nós temos que passá-lo explicitamente. Mas com o `bind` outros argumentos adicionais serão automaticamente encaminhados.
```
